DOC rework persistence user guide (#255)
skops-dev/skops: docs/persistence.rst

    untrusted source, or that could have been tampered with.

In contrast to ``pickle``, the :func:`skops.io.dump` and :func:`skops.io.load`
functions have a more limited scope, while preventing users from running
arbitrary code or loading unknown and malicious objects.

When loading a file, :func:`skops.io.load`/:func:`skops.io.loads` will read

> traverse the input and check for known and unknown types, and will only

Reviewer: ```suggestion
traverse the input, check for known and unknown types, and will only
```

Also: "will read traverse": should be "read" or "traverse", not both

.. code:: python

>     from xgboost.sklearn import XGBClassifier

Reviewer: Maybe mention above that users need xgboost to run this?

.. code:: python

    from skops.io import get_untrusted_types
    unknown_types = get_untrusted_types(file="my-model.skops")
    print(unknown_types)
    ['numpy.float64', 'numpy.int64', 'sklearn.metrics._scorer._passthrough_scorer',
    'xgboost.core.Booster', 'xgboost.sklearn.XGBClassifier']

Note that everything in the above list is safe to load. We already have many
types included as trusted by default, and some of the above values might be
added to that list in the future.

Once you check the list and you validate that everything in the list is safe,
you can load the file with ``trusted=unknown_types``:

.. code:: python

    loaded = load("my-model.skops", trusted=unknown_types)

At the moment, we support the vast majority of sklearn estimators. This
includes complex use cases such as :class:`sklearn.pipeline.Pipeline`,
:class:`sklearn.model_selection.GridSearchCV`, classes using objects defined in

> Cython such such as :class:`sklearn.tree.DecisionTreeClassifier`, and more. If

Reviewer: ```suggestion
Cython such as :class:`sklearn.tree.DecisionTreeClassifier`, and more. If
```

arrays, dtypes, random generators, and ufuncs.

Apart from this core, we plan to support machine learning libraries commonly
used be the community. So far, we have tested the following libraries:

- `LightGBM <https://lightgbm.readthedocs.io/>`_ (scikit-learn API)
- `XGBoost <https://xgboost.readthedocs.io/en/stable/>`_ (scikit-learn API)
- `CatBoost <https://catboost.ai/en/docs/>`_

If you run into a problem using any of the mentioned libraries, this could mean
there is a bug in skops. Please open an issue on `our issue tracker
<https://github.com/skops-dev/skops/issues>`__ (but please check first if a
corresponding issue already exists).

Roadmap
-------
There needs to be more testing to harden the loader and make sure we don't run
arbitrary code when it's not intended. However, the safety mechanisms already
in place should prevent most cases of abuse.

> At the moment most persisting and loading arbitrary C extension types is not
> possible unless a python object wraps around them and handles persistance and

Reviewer: ```suggestion
At the moment, persisting and loading arbitrary C extension types is not
possible, unless a python object wraps around them and handles persistence and
```

More readable IMO
